### PR TITLE
[SPARK-35886][SQL] PromotePrecision should not overwrite genCode

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch fixes `PromotePrecision` where it overwrites `genCode` where subexpression elimination should happen. 

### Why are the changes needed?

`PromotePrecision` overwrites `genCode` where subexpression elimination should happen. So if it is most top expression of a subexpression, it is never replaced.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan @wangyum @AngersZhuuuu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalExpressions.scala [hunk 1]
```diff
@@ -111,9 +111,8 @@ case class PromotePrecision(child: Expression) extends UnaryExpression {
   override def dataType: DataType = child.dataType
   override def eval(input: InternalRow): Any = child.eval(input)
   /** Just a simple pass-through for code generation. */
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalExpressions.scala:113]: Do we need to remove this obsolete comment? Now, we don't have `path-through from encode to child.genCode`

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 2]
```diff
@@ -2914,6 +2914,21 @@ class DataFrameSuite extends QueryTest
     val df2 = (1 to 10).toDF()
     assert(df2.isLocal)
   }
+
+  test("SPARK-35886: PromotePrecision should be subexpr replaced") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2918]: Can we have a test coverage in `catalyst` module?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalExpressions.scala (hunk 1 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalExpressions.scala:113]: Ok.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 2 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2918]: Ok, let me add one.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalExpressions.scala [hunk 3]
```diff
@@ -111,9 +111,8 @@ case class PromotePrecision(child: Expression) extends UnaryExpression {
...
-  override def genCode(ctx: CodegenContext): ExprCode = child.genCode(ctx)
   override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode =
-    ev.copy(EmptyBlock)
+    child.genCode(ctx)
```
Comment (AngersZhuuuu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalExpressions.scala:115]: Looks like it's the root cause, should I close my PR?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalExpressions.scala:115]: Yea, thanks! I added you as co-author.

**Review by dongjoon-hyun — APPROVED** [review]

**Review by AngersZhuuuu — APPROVED** [review]

**Comment by viirya** [comment]
Thanks @dongjoon-hyun and @AngersZhuuuu. Merging to master, 3.1, 3.0.

**Comment by cloud-fan** [comment]
late LGTM

**Pull request closed without merging**